core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java

```diff
@@ -0,0 +1,188 @@
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.client.BoundedStream;
+import alluxio.client.Locatable;
+import alluxio.client.PositionedReadable;
+import alluxio.client.Seekable;
+import alluxio.client.block.AlluxioBlockStore;
+import alluxio.client.block.BlockStoreContext;
+import alluxio.client.block.BlockWorkerClient;
+import alluxio.client.file.options.InStreamOptions;
+import alluxio.exception.AlluxioException;
+import alluxio.util.network.NetworkAddressUtils;
+import alluxio.wire.LockBlockResult;
+import alluxio.wire.WorkerNetAddress;
+import alluxio.worker.block.io.LocalFileBlockReader;
+
+import com.google.common.io.Closer;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetSocketAddress;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Provides a stream API to read a block from Alluxio. An instance extending this class can be
+ * obtained by calling {@link AlluxioBlockStore#getInStream}. Multiple
+ * {@link BlockInStream}s can be opened for a block.
+ *
+ * This class provides the same methods as a Java {@link InputStream} with additional methods from
+ * Alluxio Stream interfaces.
+ */
+@NotThreadSafe
+public final class BlockInStream extends FilterInputStream implements BoundedStream, Seekable,
+    PositionedReadable, Locatable {
+  /** Helper to manage closeables. */
+  private final Closer mCloser;
+  private final BlockWorkerClient mBlockWorkerClient;
+  private final long mBlockId;
+  private final boolean mLocal;
+  private final PacketInStream mInputStream;
+
+  /**
+   * Creates an instance of {@link BlockInStream}.
+   *
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param workerNetAddress the worker network address
+   * @param context the block store context
+   * @param options the options
+   * @throws IOException if it fails to create an instance
+   * @return the {@link BlockInStream} created
+   */
+  public static BlockInStream createLocalBlockInStream(long blockId, long blockSize,
+      WorkerNetAddress workerNetAddress, BlockStoreContext context, InStreamOptions options)
+      throws IOException {
+    Closer closer = Closer.create();
+
+    try {
+      BlockWorkerClient blockWorkerClient =
+          closer.register(context.createWorkerClient(workerNetAddress));
+      LockBlockResult lockBlockResult = blockWorkerClient.lockBlock(blockId);
+      LocalFileBlockReader localFileBlockReader =
+          new LocalFileBlockReader(lockBlockResult.getBlockPath());
+      PacketReader.Factory factory = new LocalFilePacketReader.Factory(localFileBlockReader);
```

Comment:
This is actually required (I have thought of a better way to do this) because BlockInStream needs to create PacketReader on demand. In other words, one BlockInStream can create many PacketReaders. This is different from BlockOutStream